src/repr/src/row.rs

```diff
@@ -1848,13 +2141,214 @@ mod tests {
+            (Datum::Int32(-1), true, Datum::Int32(1), true, "[-1,1]"),
+            (Datum::Int32(-1), false, Datum::Int32(1), false, "(-1,1)"),
+            (Datum::Int32(-1), false, Datum::Int32(1), true, "(-1,1]"),
+            // Empty + point ranges
+            (Datum::Int32(0), true, Datum::Int32(0), false, "[0,0)"),
+            (Datum::Int32(0), true, Datum::Int32(0), true, "[0,0]"),
+            (Datum::Int32(0), false, Datum::Int32(0), false, "(0,0)"),
+            (Datum::Int32(0), false, Datum::Int32(0), true, "(0,0]"),
+            // Unbounded upper
+            (Datum::Int32(1), true, Datum::Null, false, "[1,)"),
+            (Datum::Int32(1), true, Datum::Null, true, "[1,]"),
+            (Datum::Int32(1), false, Datum::Null, false, "(1,)"),
+            (Datum::Int32(1), false, Datum::Null, true, "(1,]"),
+        ] {
+            packer
+                .push_range(lower, lower_inc, upper, upper_inc)
+                .unwrap();
+            let r = arena.make_datum(|packer| {
+                packer
+                    .push_range_with(
+                        RangeBoundDesc {
+                            inf: matches!(lower, Datum::Null),
+                            inc: lower_inc,
+                        },
+                        RangeBoundDesc {
+                            inf: matches!(upper, Datum::Null),
+                            inc: upper_inc,
+                        },
+                        |row| {
+                            if !matches!(lower, Datum::Null) {
+                                row.push(lower);
+                            }
+                            if !matches!(upper, Datum::Null) {
+                                row.push(upper);
+                            }
+                        },
+                    )
+                    .unwrap();
+            });
+            datum_vec.push((r, s));
+        }
+
+        let empty_range = arena.make_datum(|packer| {
+            packer.push_empty_range();
+        });
+        let range_0 = arena.make_datum(|packer| {
+            packer.push(Datum::Range(Range { inner: None }));
+        });
+
+        test_ranges_inner(empty_range, &range_0, &"empty");
+
+        let mut last_seen = empty_range;
+
+        for (row_value, (vec_value, string_rep)) in row.iter().zip_eq(datum_vec.iter()) {
+            // Validates Ord implementation; relies on input being hand-sorted.
+            assert!(last_seen < row_value, "{last_seen} < {row_value}");
+            test_ranges_inner(row_value, vec_value, string_rep);
+            last_seen = row_value;
+        }
+
+        // Test null handling
+        for (lower, upper, s) in [
+            // This essentially tests the encoding we'll need for non-partition
+            // reclocking data.
+            (None, Some(Datum::Null), "(,null)"),
+            (Some((Datum::Null, true)), None, "[null,)"),
+            (Some((Datum::Null, false)), None, "(null,)"),
```

Comment:
I'm confused about this! PostgreSQL doesn't support `null` as a range boundary. It seems like we shouldn't either? Do we need this for reclocking?

Reply:
Yes, unfortunately we need this for reclocking, or rather, this is the most sensible value to use for `PartitionId::None` in reclocking. The value never gets surfaced or returned to users, so permitting it "back here" seemed fine. We could use some other magical value, but this seemed the most semantic.

Reply:
Oh, weird, ok. My specific concern is that `NULL` is used to mean "infinity" in certain contexts in PostgreSQL ranges

```
benesch=# select int4range(NULL, 1);
 int4range 
-----------
 (,1)
(1 row)
```

but as long as this is internal only it's probably fine! Why doesn't this ever get exposed to users? Does the reclocking relation just not expose the `partition` column for source types that don't have meaningful partitions?

Reply:
Dug into this and it's possible we don't need this null-supporting feature. I'll add the panics on `Datum::Null` back in and can remove them if native timestamp reclocking turns out to need them.